# Ebay Car Sales Data

## Overview of the data

In [1]:
import numpy as np
import pandas as pd

In [2]:
autos = pd.read_csv('./dataset/autos.csv', encoding = 'Latin_1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

A neat feature of jupyter notebook is its ability to render the first few and last few values of any pandas object.

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

* The dataset contains 20 columns, most of which are strings.
* Some columns have null values, but none have more than ~20% null values. (how did they find this?)
* The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

## Data cleaning

### Cleaning Column Names
Converting the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
import re

new_columns = []

def clean_col(col):
    col = col.replace("yearOfRegistartion","registration_year")
    col = col.replace("monthOfRegistration", "registration_month")
    col = col.replace("yearOfRegistration", "registration_year")
    col = col.replace("notRepairedDamage", "unrepaired_damage")
    col = col.replace("dateCreated", "ad_created")
    col = re.sub(r'(?<!^)(?=[A-Z])', '_', col).lower()
    return col


for names in autos.columns:
    cleaned_col = clean_col(names)
    new_columns.append(cleaned_col)
    
autos.columns = new_columns

In [7]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

### Initial Exploration and Cleaning

Some basic data exploration to determine what other cleaning tasks need to be done. 

Initially will look for:
* Text columns where all or almost all values are the same. 
* Examples of numeric data stored as text which can be cleaned and converted.

In [8]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [9]:
print(autos.isnull().sum())

date_crawled             0
name                     0
seller                   0
offer_type               0
price                    0
abtest                   0
vehicle_type          5095
registration_year        0
gearbox               2680
power_p_s                0
model                 2758
odometer                 0
registration_month       0
fuel_type             4482
brand                    0
unrepaired_damage     9829
ad_created               0
nr_of_pictures           0
postal_code              0
last_seen                0
dtype: int64


In [10]:
autos[autos['registration_month'] < 1].describe()

registration_year    power_p_s  registration_month  nr_of_pictures  \
count        5075.000000  5075.000000              5075.0          5075.0   
mean         2012.730049    66.869754                 0.0             0.0   
std           285.652252    84.319912                 0.0             0.0   
min          1000.000000     0.000000                 0.0             0.0   
25%          1997.000000     0.000000                 0.0             0.0   
50%          2000.000000    60.000000                 0.0             0.0   
75%          2006.000000   113.000000                 0.0             0.0   
max          9999.000000  1998.000000                 0.0             0.0   

        postal_code  
count   5075.000000  
mean   48120.804335  
std    25955.661438  
min     1067.000000  
25%    26906.000000  
50%    46325.000000  
75%    66613.000000  
max    99986.000000

5,075 cars have a registration month of 0. This needs to be adjusted but might not be that useful anyway.

Numeric data stored as text that needs to be cleaned:
* `price`
* `kilometer`


In [11]:
print(autos["price"].dtype)
print(autos["price"].unique())

object
['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']


In [12]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)

autos.rename({"price": "price_$"}, axis=1, inplace=True)

print(autos["price_$"].dtype)
print(autos["price_$"].unique())

int64
[ 5000  8500  8990 ...   385 22200 16995]


In [13]:
autos["price_$"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64

In [14]:
print(autos["odometer"].dtype)
print(autos["odometer"].unique())

object
['150,000km' '70,000km' '50,000km' '80,000km' '10,000km' '30,000km'
 '125,000km' '90,000km' '20,000km' '60,000km' '5,000km' '100,000km'
 '40,000km']


In [15]:
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

print(autos["odometer_km"].dtype)
print(autos["odometer_km"].unique())

int64
[150000  70000  50000  80000  10000  30000 125000  90000  20000  60000
   5000 100000  40000]


### Exploring the Odometer and Price Columns

Continue exploring the `price` and `odometer_km` data, specifically data that doesn't look right.

Looking for any values that look unrealistically high or low (outliers) that we might want to remove.

In [16]:
print(autos["price_$"].unique().shape)
print('\n')
print(autos["price_$"].describe())

(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64


In [17]:
autos["price_$"].value_counts().head()

0       1421
500      781
1500     734
2500     643
1000     639
Name: price_$, dtype: int64

In [18]:
autos["price_$"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_$, dtype: int64

In [19]:
autos["price_$"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_$, dtype: int64

In [20]:
autos[autos["price_$"] == 3890000]

date_crawled         name  seller offer_type  price_$ abtest  \
47634  2016-04-04 21:25:21  Ferrari_FXX  privat    Angebot  3890000   test   

      vehicle_type  registration_year gearbox  power_p_s model  odometer_km  \
47634        coupe               2006     NaN        799   NaN         5000   

       registration_month fuel_type           brand unrepaired_damage  \
47634                   7       NaN  sonstige_autos              nein   

                ad_created  nr_of_pictures  postal_code            last_seen  
47634  2016-04-04 00:00:00               0        60313  2016-04-05 12:07:37

Removing outliers:
* 1,421 with a price of $0
* Cars above (and including) the price of  999,990


In [21]:
autos = autos.drop(autos[autos["price_$"] == 0].index)
autos = autos.drop(autos[autos["price_$"] >= 999990 ].index)


autos["price_$"].value_counts().sort_index(ascending=True).head()

1    156
2      3
3      1
5      2
8      1
Name: price_$, dtype: int64

In [22]:
autos["price_$"].value_counts().sort_index(ascending=False).head(10)

350000    1
345000    1
299000    1
295000    1
265000    1
259000    1
250000    1
220000    1
198000    1
197000    1
Name: price_$, dtype: int64

In [23]:
print(autos["odometer_km"].unique().shape)
print('\n')
print(autos["odometer_km"].describe())

(13,)


count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [24]:
autos["odometer_km"].value_counts().head()

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
Name: odometer_km, dtype: int64

Not removed any autos based on `odometer_km` values since no outliers.

### Exploring the data columns

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: (when the crawler saw this ad last online) added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. 

First understanding how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [25]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The first 10 characters represent the day (e.g. `2016-03-12`). 

In [26]:
print(autos['date_crawled'].str[:10])

autos['date_crawled'] = autos['date_crawled'].str[:10]

autos['date_crawled']

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48565, dtype: object


0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48565, dtype: object

In [27]:
print(autos['last_seen'].str[:10])

autos['last_seen'] = autos['last_seen'].str[:10]

autos['last_seen']

0        2016-04-06
1        2016-04-06
2        2016-04-06
3        2016-03-15
4        2016-04-01
            ...    
49995    2016-04-01
49996    2016-04-02
49997    2016-04-04
49998    2016-04-05
49999    2016-04-06
Name: last_seen, Length: 48565, dtype: object


0        2016-04-06
1        2016-04-06
2        2016-04-06
3        2016-03-15
4        2016-04-01
            ...    
49995    2016-04-01
49996    2016-04-02
49997    2016-04-04
49998    2016-04-05
49999    2016-04-06
Name: last_seen, Length: 48565, dtype: object

In [28]:
print(autos['ad_created'].str[:10])

autos['ad_created'] = autos['ad_created'].str[:10]

autos['ad_created']

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-13
Name: ad_created, Length: 48565, dtype: object


0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-13
Name: ad_created, Length: 48565, dtype: object

To understand the date range, we can extract just the date values, use `Series.value_counts()` to generate a distribution, and then sort by the index.

In [29]:
autos['date_crawled'].value_counts(normalize=True, dropna=False)

2016-04-03    0.038608
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-12    0.036920
2016-03-14    0.036549
2016-04-04    0.036487
2016-03-07    0.036014
2016-04-02    0.035478
2016-03-28    0.034860
2016-03-19    0.034778
2016-03-15    0.034284
2016-03-29    0.034099
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-22    0.032987
2016-03-11    0.032575
2016-03-23    0.032225
2016-03-26    0.032204
2016-03-10    0.032184
2016-03-31    0.031834
2016-03-17    0.031628
2016-03-25    0.031607
2016-03-27    0.031092
2016-03-16    0.029610
2016-03-24    0.029342
2016-03-05    0.025327
2016-03-13    0.015670
2016-03-06    0.014043
2016-04-05    0.013096
2016-03-18    0.012911
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The most data extracted was on 3rd April 2016 at 3.86%, followed by 20th March 2016 at 3.79% of all the data. 

In [30]:
autos['ad_created'].value_counts(normalize=True, dropna=False)

2016-04-03    0.038855
2016-03-20    0.037949
2016-03-21    0.037579
2016-04-04    0.036858
2016-03-12    0.036755
                ...   
2016-01-16    0.000021
2015-09-09    0.000021
2016-01-03    0.000021
2016-01-13    0.000021
2016-02-07    0.000021
Name: ad_created, Length: 76, dtype: float64

The most ads extracted were on 3rd April 2016 at 3.89%, followed by 20th March 2016 at 3.79% of all the data. Exactly the same dates as `date_crawled`.

In [31]:
autos['last_seen'].value_counts(normalize=True, dropna=False)

2016-04-06    0.221806
2016-04-07    0.131947
2016-04-05    0.124761
2016-03-17    0.028086
2016-04-03    0.025203
2016-04-02    0.024915
2016-03-30    0.024771
2016-04-04    0.024483
2016-03-12    0.023783
2016-03-31    0.023783
2016-04-01    0.022794
2016-03-29    0.022341
2016-03-22    0.021373
2016-03-28    0.020859
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-23    0.018532
2016-03-26    0.016802
2016-03-16    0.016452
2016-03-15    0.015876
2016-03-19    0.015834
2016-03-27    0.015649
2016-03-14    0.012602
2016-03-11    0.012375
2016-03-10    0.010666
2016-03-09    0.009595
2016-03-13    0.008895
2016-03-08    0.007413
2016-03-18    0.007351
2016-03-07    0.005395
2016-03-06    0.004324
2016-03-05    0.001071
Name: last_seen, dtype: float64

The most popular date that the crawler last the ads online was on 6th April 2016 at 22.2%.

In [32]:
autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The first date, for this dataset, that data was extracted/crawled was 5th March 2016 and the last date was 7th April 2016.

In [33]:
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

The earliest date, for this dataset, that an ad was created was 11th June 2011 and the last date was 7th April 2016 (same as `date_crawled`).

In [34]:
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

The earliest date, for this dataset, that the crawler saw this add online was 11th June 2016 and the last date was 7th April 2016.

### Dealing with Incorrect Registration Year Data

In [35]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

`registration_year` has questionable data (min = 1000) also max = 9999

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. 

Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Will count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

Removing cars registered before 1930 as the list of cars that are registered before 1930 don't appear to be legitimate.

In [36]:
autos[(autos["registration_year"] > 2016)]

date_crawled                                               name  seller  \
10      2016-03-15                       VW_Golf_Tuning_in_siber/grau  privat   
55      2016-03-07                     Mercedes_E320_AMG_zu_Tauschen!  privat   
65      2016-04-04                      Ford_Fiesta_zum_ausschlachten  privat   
68      2016-04-03  Mini_cooper_s_clubman_/vollausstattung_/_Navi/...  privat   
84      2016-03-27                                     Renault_twingo  privat   
...            ...                                                ...     ...   
49770   2016-03-15                     VW_Polo_6n_Tuev_Neu!__1.6_75PS  privat   
49796   2016-03-09                        Opel_corsa_1.4_zu_verkaufen  privat   
49841   2016-03-11                                  Passat_abzugeben.  privat   
49910   2016-04-03  Schoener_fast_neuer_Opel_Mokka_in_Zell_Mosel_m...  privat   
49935   2016-04-01  Mercedes_A_klasse_angemeldet_mit_Tuef_und_Auto...  privat   

      offer_type  price_$   abtest vehicle_type  registration_year    gearbox  \
10       Angebot      999     test          NaN               2017    manuell   
55       Angebot        1     test          NaN               2017  automatik   
65       Angebot      250  control          NaN               2017    manuell   
68       Angebot    10990     test          NaN               2017    manuell   
84       Angebot      900  control          NaN               2018        NaN   
...          ...      ...      ...          ...                ...        ...   
49770    Angebot      999  control          NaN               2018    manuell   
49796    Angebot     4500     test          NaN               2017    manuell   
49841    Angebot      600     test          NaN               2017    manuell   
49910    Angebot    22200     test          NaN               9000  automatik   
49935    Angebot      800     test          NaN               2017  automatik   

       power_p_s     model  odometer_km  registration_month fuel_type  \
10            90       NaN       150000                   4    benzin   
55           224  e_klasse       125000                   7    benzin   
65            65    fiesta       125000                   9    benzin   
68           174   clubman       100000                   0       NaN   
84            60    twingo       150000                   0       NaN   
...          ...       ...          ...                 ...       ...   
49770         75      polo       150000                  12    benzin   
49796         90     corsa        70000                   7    benzin   
49841        101    passat       150000                   7       NaN   
49910        140    andere        10000                   3    benzin   
49935        101  a_klasse       150000                   9       NaN   

               brand unrepaired_damage  ad_created  nr_of_pictures  \
10        volkswagen              nein  2016-03-14               0   
55     mercedes_benz              nein  2016-03-06               0   
65              ford               NaN  2016-04-04               0   
68              mini              nein  2016-04-03               0   
84           renault               NaN  2016-03-27               0   
...              ...               ...         ...             ...   
49770     volkswagen              nein  2016-03-15               0   
49796           opel              nein  2016-03-09               0   
49841     volkswagen               NaN  2016-03-11               0   
49910           opel               NaN  2016-04-03               0   
49935  mercedes_benz              nein  2016-04-01               0   

       postal_code   last_seen  
10           86157  2016-04-07  
55           22111  2016-03-08  
65           65606  2016-04-05  
68           83135  2016-04-05  
84           40589  2016-04-05  
...            ...         ...  
49770        24321  2016-04-06  
49796        88433  2016-03-17  
49841        53804  2016-03-11  
49910       

In [37]:
autos[(autos["registration_year"] < 1930)]

date_crawled                                   name  seller offer_type  \
3679    2016-04-04                             Suche_Auto  privat    Angebot   
10556   2016-04-01                             UNFAL_Auto  privat    Angebot   
21416   2016-03-12                Essex_super_six__Ford_A  privat    Angebot   
22101   2016-03-09                             BMW_Andere  privat    Angebot   
22316   2016-03-29  VW_Kaefer.__Zwei_zum_Preis_von_einem.  privat    Angebot   
22659   2016-03-14                           Opel_Corsa_B  privat    Angebot   
24511   2016-03-17            Trabant__wartburg__Ostalgie  privat    Angebot   
28693   2016-03-22                         Renault_Twingo  privat    Angebot   
30781   2016-03-25       Opel_Calibra_V6_DTM_Bausatz_1:24  privat    Angebot   
32585   2016-04-02                             UNFAL_Auto  privat    Angebot   
45157   2016-03-11                             Motorhaube  privat    Angebot   
49283   2016-03-15                             Citroen_HY  privat    Angebot   

       price_$   abtest vehicle_type  registration_year  gearbox  power_p_s  \
3679         1     test          NaN               1910      NaN          0   
10556      450  control          NaN               1800      NaN       1800   
21416    16500  control       cabrio               1927  manuell         40   
22101    11500     test       cabrio               1929  manuell         15   
22316     1500  control          NaN               1000  manuell          0   
22659      500     test          NaN               1910      NaN          0   
24511      490  control          NaN               1111      NaN          0   
28693      599  control   kleinwagen               1910  manuell          0   
30781       30     test          NaN               1910      NaN          0   
32585      450  control          NaN               1800      NaN       1800   
45157       15  control          NaN               1910      NaN          0   
49283     7750  control          NaN               1001      NaN          0   

         model  odometer_km  registration_month fuel_type           brand  \
3679       NaN         5000                   0       NaN  sonstige_autos   
10556      NaN         5000                   2       NaN      mitsubishi   
21416   andere         5000                   5    benzin            ford   
22101   andere         5000                   1       NaN             bmw   
22316   kaefer         5000                   0    benzin      volkswagen   
22659    corsa       150000                   0       NaN            opel   
24511      NaN         5000                   0       NaN         trabant   
28693      NaN         5000                   0    benzin         renault   
30781  calibra       100000                   0       NaN            opel   
32585      NaN         5000                   2       NaN      mitsubishi   
45157      NaN         5000                   0       NaN         trabant   
49283   andere         5000                   0       NaN         citroen   

      unrepaired_damage  ad_created  nr_of_pictures  postal_code   last_seen  
3679                NaN  2016-04-04               0        40239  2016-04-04  
10556              nein  2016-04-01               0        63322  2016-04-01  
21416               NaN  2016-03-12               0        74821  2016-03-15  
22101                ja  2016-03-09               0        70569  2016-04-07  
22316               NaN  2016-03-29               0        48324  2016-03-31  
22659               NaN  2016-03-14               0        52393  2016-04-03  
24511               NaN  2016-03-17               0        16818  2016-04-07  
28693               NaN  2016-03-22               0        70376  2016-04-06  
30781               NaN  2016-03-25               0        47638  2016-03-26  
32585              nein  2016-04-02               0        63322  2016-04-04  
45157               NaN  2016-03-11               0        90491  

In [38]:
print(autos["registration_year"].value_counts(normalize=True))

reg_year = autos[(autos["registration_year"].between(1929, 2017))]
print(reg_year["registration_year"].value_counts(normalize=True))

autos = autos.drop(autos[autos["registration_year"].between(1929, 2017)].index)

2000    0.064985
2005    0.060455
1999    0.059652
2004    0.055657
2003    0.055575
          ...   
9000    0.000021
1948    0.000021
1000    0.000021
1931    0.000021
1952    0.000021
Name: registration_year, Length: 95, dtype: float64
2000    0.065658
2005    0.061081
1999    0.060270
2004    0.056234
2003    0.056151
          ...   
1948    0.000021
1938    0.000021
1939    0.000021
1943    0.000021
1952    0.000021
Name: registration_year, Length: 77, dtype: float64


In [39]:
print(reg_year["registration_year"].value_counts(normalize=True))

2000    0.065658
2005    0.061081
1999    0.060270
2004    0.056234
2003    0.056151
          ...   
1948    0.000021
1938    0.000021
1939    0.000021
1943    0.000021
1952    0.000021
Name: registration_year, Length: 77, dtype: float64


Most cars are registered in 2000 (6.6%) followed by 2005 (6.1%) and 1999 (6.0%)

### Exploring Price by Brand

In [40]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.232932
opel              0.138554
bmw               0.086345
mercedes_benz     0.084337
renault           0.068273
audi              0.058233
ford              0.052209
seat              0.050201
fiat              0.040161
sonstige_autos    0.018072
honda             0.016064
citroen           0.016064
peugeot           0.014056
mitsubishi        0.014056
smart             0.012048
mazda             0.012048
nissan            0.010040
chevrolet         0.008032
suzuki            0.008032
hyundai           0.006024
kia               0.006024
skoda             0.006024
volvo             0.004016
alfa_romeo        0.004016
subaru            0.004016
trabant           0.004016
toyota            0.004016
lada              0.004016
daihatsu          0.004016
jaguar            0.002008
daewoo            0.002008
saab              0.002008
lancia            0.002008
porsche           0.002008
jeep              0.002008
dacia             0.002008
Name: brand, dtype: float64

The top 3 listed car brands are VW (23%), Opel (14%), BMW (9%).

Will be aggregating on the top 10 cars (excluding 'sonstige_autos' since this is 'other cars') to identify the average price of these brands and order by price. 

In [41]:
autos[autos["brand"] == 'sonstige_autos']

date_crawled                                               name  seller  \
2051    2016-03-21                                     Wartburg_353_W  privat   
3679    2016-04-04                                         Suche_Auto  privat   
4549    2016-04-01                                         Kompressor  privat   
20732   2016-03-18                                      Auto_gesucht!  privat   
25003   2016-03-06              Reo_Vorkriegs_Oldtimer_Rennwagen_1928  privat   
25318   2016-03-05                                       Auto_gesucht  privat   
34099   2016-04-03  Cadillac_Deville_Cabrio_Bj_68_H_Kennzeichen_gu...  privat   
40583   2016-04-01                                  2_Autos_zum_tauch  privat   
42079   2016-03-17                      APE_50___deutsche_Papiere_!!!  privat   

      offer_type  price_$   abtest vehicle_type  registration_year    gearbox  \
2051     Angebot     1300     test          NaN               2018    manuell   
3679     Angebot        1     test          NaN               1910        NaN   
4549     Angebot     1600     test          NaN               4100        NaN   
20732    Angebot     1200  control          NaN               2018    manuell   
25003    Angebot     6500  control          NaN               8888        NaN   
25318    Angebot      600  control          NaN               2018        NaN   
34099    Angebot    20950  control          NaN               2018  automatik   
40583    Angebot      900  control          NaN               2018    manuell   
42079    Angebot      500  control          NaN               4800    manuell   

       power_p_s model  odometer_km  registration_month fuel_type  \
2051           0   NaN       100000                   1       NaN   
3679           0   NaN         5000                   0       NaN   
4549           0   NaN         5000                   0       NaN   
20732          0   NaN       150000                   0    benzin   
25003          0   NaN        10000                   0       NaN   
25318          0   NaN       100000                   0       NaN   
34099          0   NaN       150000                   1       NaN   
40583         75   NaN       150000                   6       NaN   
42079          3   NaN         5000                  11       NaN   

                brand unrepaired_damage  ad_created  nr_of_pictures  \
2051   sonstige_autos               NaN  2016-03-21               0   
3679   sonstige_autos               NaN  2016-04-04               0   
4549   sonstige_autos               NaN  2016-04-01               0   
20732  sonstige_autos              nein  2016-03-18               0   
25003  sonstige_autos               NaN  2016-03-06               0   
25318  sonstige_autos               NaN  2016-03-05               0   
34099  sonstige_autos               NaN  2016-04-03               0   
40583  sonstige_autos               NaN  2016-04-01               0   
42079  sonstige_autos               NaN  2016-03-17               0   

       postal_code   last_seen  
2051          8606  2016-03-28  
3679         40239  2016-04-04  
4549         67686  2016-04-05  
20732        66802  2016-03-18  
25003        55262  2016-03-30  
25318        33178  2016-03-16  
34099        78476  2016-04-06  
40583         2788  2016-04-01  
42079        48653  2016-03-17

In [42]:
autos[autos["name"].str.contains('auto', case=False)]

date_crawled                                               name  seller  \
1551    2016-03-08            Verkaufe_hier_meine_Auto_vw_Passat_2007  privat   
3679    2016-04-04                                         Suche_Auto  privat   
4213    2016-03-21                       Audi_A4_Automátik_und_Diésel  privat   
7272    2016-03-21                      Auto_fahrbereit_gutem_zustand  privat   
10556   2016-04-01                                         UNFAL_Auto  privat   
13089   2016-03-10           Toyota_Corolla_Automatik_mit_2Jahren_TÜV  privat   
13835   2016-03-24          Mercedes_E_200_Automatik_74060km_Tuev_neu  privat   
14549   2016-03-12               Seat_Arosa_1_0_City_Auto_TÜV_ASU_NEU  privat   
15176   2016-03-24                    Dringend_ein_neues_Auto_gesucht  privat   
17881   2016-03-30  LPG_Autogas_BMW_320i_M_Paket_6_Zylinder_TÜV_Ne...  privat   
18345   2016-03-11      Mitsubischi_colt_zc3_tuev_2/18_anfaenger_auto  privat   
19523   2016-04-02   BMW_320i_Coupé_Automatik_TÜV_Neu_mit_Sonnenrollo  privat   
20730   2016-03-15  VW_Polo_6N___Bj._1997___Neuer_TÜV___204.000_km...  privat   
20732   2016-03-18                                      Auto_gesucht!  privat   
22127   2016-03-12                               Suche_auto_auf_raten  privat   
23314   2016-03-26  Gepflegter_VW_Touran_1.9TDI_Standheizung_Klima...  privat   
25318   2016-03-05                                       Auto_gesucht  privat   
25445   2016-03-08  Suzuki_Swift_1.6_Sport__*sehr_sparsam*Autogas*...  privat   
26406   2016-03-20  Citroen_C3_Automatik___fette_Alus_+_LED_Rueckl...  privat   
27618   2016-03-29  Golf_1_75ps_5911km_Original_Automatik_einer_de...  privat   
28120   2016-03-07           Auto_zu_vermieten_Kurz_Langzeit_30_400_€  privat   
32585   2016-04-02                                         UNFAL_Auto  privat   
36165   2016-03-08                Golf_4_Automatik__1_6__Tuev_03/2018  privat   
36809   2016-03-31                                  Auto_zu_verkaufen  privat   
37519   2016-03-23          A6_2.5_tdi_org.115000km._leder._Automatik  privat   
37966   2016-04-03  Ich_habe_das_Auto_von_VW_gekauft._Und_hatte_no...  privat   
40583   2016-04-01                                  2_Autos_zum_tauch  privat   
42022   2016-03-20                      BMW_318d_Automatik_Navigation  privat   
45879   2016-03-31                            Audi_A3_1.8_T_Automatik  privat   
46716   2016-03-31                                          Auto_Soda  privat   
47340   2016-03-13  Golf_4_Cabrio_1.6_mit_Automatikgetriebe_und_Tu...  privat   
48249   2016-03-21  Mercedes_Benz_Viano_Trend_Diesel_Automatik_kom...  privat   

      offer_type  price_$   abtest vehicle_type  registration_year    gearbox  \
1551     Angebot     3200     test          NaN               2018    manuell   
3679     Angebot        1     test          NaN               1910        NaN   
4213     Angebot     1499  control          NaN               2018  automatik   
7272     Angebot     1500     test          NaN               2018    manuell   
10556    Angebot      450  control          NaN               1800        NaN   
13089    Angebot     1899  control          NaN               2018  automatik   
13835    Angebot    10990     test          NaN               2018  automatik   
14549    Angebot     1525     test          NaN               2018    manuell   
15176    Angebot     3500     test          NaN               2018        NaN   
17881    Angebot     2750  control          NaN               2018    manuell   
18345    Angebot     2100  control          NaN               2018    manuell   
19523    Angebot     1000     test          NaN               2018  automatik   
20730    Angebot      870  control          NaN               2018    manuell   
20732    Angebot     1200  control          NaN               2018    manuell   
22127    Angebot      500  control          NaN               2018        NaN   
23314    Angebot     5999     test  

Have noticed that there are some listings where people are looking for cars and not selling. Will remove these.

In [43]:
autos = autos.drop([3679, 15176, 20732, 22127, 25318, 40583], axis=0)

autos[autos["name"].str.contains('auto', case=False)]

date_crawled                                               name  seller  \
1551    2016-03-08            Verkaufe_hier_meine_Auto_vw_Passat_2007  privat   
4213    2016-03-21                       Audi_A4_Automátik_und_Diésel  privat   
7272    2016-03-21                      Auto_fahrbereit_gutem_zustand  privat   
10556   2016-04-01                                         UNFAL_Auto  privat   
13089   2016-03-10           Toyota_Corolla_Automatik_mit_2Jahren_TÜV  privat   
13835   2016-03-24          Mercedes_E_200_Automatik_74060km_Tuev_neu  privat   
14549   2016-03-12               Seat_Arosa_1_0_City_Auto_TÜV_ASU_NEU  privat   
17881   2016-03-30  LPG_Autogas_BMW_320i_M_Paket_6_Zylinder_TÜV_Ne...  privat   
18345   2016-03-11      Mitsubischi_colt_zc3_tuev_2/18_anfaenger_auto  privat   
19523   2016-04-02   BMW_320i_Coupé_Automatik_TÜV_Neu_mit_Sonnenrollo  privat   
20730   2016-03-15  VW_Polo_6N___Bj._1997___Neuer_TÜV___204.000_km...  privat   
23314   2016-03-26  Gepflegter_VW_Touran_1.9TDI_Standheizung_Klima...  privat   
25445   2016-03-08  Suzuki_Swift_1.6_Sport__*sehr_sparsam*Autogas*...  privat   
26406   2016-03-20  Citroen_C3_Automatik___fette_Alus_+_LED_Rueckl...  privat   
27618   2016-03-29  Golf_1_75ps_5911km_Original_Automatik_einer_de...  privat   
28120   2016-03-07           Auto_zu_vermieten_Kurz_Langzeit_30_400_€  privat   
32585   2016-04-02                                         UNFAL_Auto  privat   
36165   2016-03-08                Golf_4_Automatik__1_6__Tuev_03/2018  privat   
36809   2016-03-31                                  Auto_zu_verkaufen  privat   
37519   2016-03-23          A6_2.5_tdi_org.115000km._leder._Automatik  privat   
37966   2016-04-03  Ich_habe_das_Auto_von_VW_gekauft._Und_hatte_no...  privat   
42022   2016-03-20                      BMW_318d_Automatik_Navigation  privat   
45879   2016-03-31                            Audi_A3_1.8_T_Automatik  privat   
46716   2016-03-31                                          Auto_Soda  privat   
47340   2016-03-13  Golf_4_Cabrio_1.6_mit_Automatikgetriebe_und_Tu...  privat   
48249   2016-03-21  Mercedes_Benz_Viano_Trend_Diesel_Automatik_kom...  privat   

      offer_type  price_$   abtest vehicle_type  registration_year    gearbox  \
1551     Angebot     3200     test          NaN               2018    manuell   
4213     Angebot     1499  control          NaN               2018  automatik   
7272     Angebot     1500     test          NaN               2018    manuell   
10556    Angebot      450  control          NaN               1800        NaN   
13089    Angebot     1899  control          NaN               2018  automatik   
13835    Angebot    10990     test          NaN               2018  automatik   
14549    Angebot     1525     test          NaN               2018    manuell   
17881    Angebot     2750  control          NaN               2018    manuell   
18345    Angebot     2100  control          NaN               2018    manuell   
19523    Angebot     1000     test          NaN               2018  automatik   
20730    Angebot      870  control          NaN               2018    manuell   
23314    Angebot     5999     test          NaN               2018    manuell   
25445    Angebot     4850  control          NaN               2018  automatik   
26406    Angebot     2995     test          NaN               2018  automatik   
27618    Angebot     7999     test          NaN               5911  automatik   
28120    Angebot       30     test          NaN               2018    manuell   
32585    Angebot      450  control          NaN               1800        NaN   
36165    Angebot     2190  control          NaN               2018  automatik   
36809    Angebot     3000     test          NaN               2018    manuell   
37519    Angebot     5950     test          NaN               2018  automatik   
37966    Angebot     6200     test          NaN               2018    manuell   
42022    Angebot    13800     test  

In [44]:
autos["brand"].value_counts(normalize=True)[:6]

volkswagen       0.235772
opel             0.140244
bmw              0.087398
mercedes_benz    0.083333
renault          0.069106
audi             0.058943
Name: brand, dtype: float64

Of the top 6 most popular car brands listed, Renault cars are the cheapest cars on average and the most expensive car brand (on average) is Mercedes Benz.  

In [45]:
# Create an empty dictionary to store the results
brand_mean_prices = {}

#Create an array of the top 6 brands
autos_top6 = autos["brand"].value_counts(normalize=True)[:6].index.tolist()

#Use a for loop to iterate over the brands
for auto in autos_top6:
    selected_rows = autos[autos["brand"] == auto]
    mean = selected_rows["price_$"].mean()
    brand_mean_prices[auto] = mean
  

#sorting the dictionary by the lowest price

{k: v for k, v in sorted(
    brand_mean_prices.items(), key=lambda item: item[1]
)}

{'renault': 1597.235294117647,
 'opel': 2277.6666666666665,
 'volkswagen': 3665.0,
 'audi': 5446.103448275862,
 'bmw': 5503.3488372093025,
 'mercedes_benz': 9604.439024390244}

Calculating mean mileage (for the next exercise)

In [46]:
# Create an empty dictionary to store the results
brand_mean_mileage = {}

#Create an array of the top 6 brands
autos_top6 = autos["brand"].value_counts(normalize=True)[:6].index.tolist()

#Use a for loop to iterate over the brands
for auto in autos_top6:
    selected_rows = autos[autos["brand"] == auto]
    mean = selected_rows["odometer_km"].mean()
    brand_mean_mileage[auto] = mean
  

{k: v for k, v in sorted(
    brand_mean_mileage.items(), key=lambda item: item[1]
)}

{'mercedes_benz': 115243.90243902439,
 'renault': 122205.88235294117,
 'volkswagen': 126120.68965517242,
 'opel': 126811.59420289854,
 'bmw': 134651.16279069768,
 'audi': 141724.1379310345}

### Average mileage for the top 6 car brands

For the top 6 brands, will use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. 

In [47]:
# Dictionary to series 

bmp_series = pd.Series(brand_mean_prices)
print(bmp_series)

print('\n')

bmm_series = pd.Series(brand_mean_mileage)
print(bmm_series)


volkswagen       3665.000000
opel             2277.666667
bmw              5503.348837
mercedes_benz    9604.439024
renault          1597.235294
audi             5446.103448
dtype: float64


volkswagen       126120.689655
opel             126811.594203
bmw              134651.162791
mercedes_benz    115243.902439
renault          122205.882353
audi             141724.137931
dtype: float64


In [48]:
# Series to Dataframe

brands = pd.DataFrame(bmp_series, columns=['mean_price'])
brands

mean_price
volkswagen     3665.000000
opel           2277.666667
bmw            5503.348837
mercedes_benz  9604.439024
renault        1597.235294
audi           5446.103448

In [49]:
## Assigning bmm to above Dataframe

brands["mean_mileage"] = bmm_series
brands

mean_price   mean_mileage
volkswagen     3665.000000  126120.689655
opel           2277.666667  126811.594203
bmw            5503.348837  134651.162791
mercedes_benz  9604.439024  115243.902439
renault        1597.235294  122205.882353
audi           5446.103448  141724.137931

In [50]:
pd.options.display.float_format = '{:,.0f}'.format
brands

mean_price  mean_mileage
volkswagen          3,665       126,121
opel                2,278       126,812
bmw                 5,503       134,651
mercedes_benz       9,604       115,244
renault             1,597       122,206
audi                5,446       141,724

In [51]:
brands["value_for_money"] =  (brands["mean_price"]) / (brands["mean_mileage"])
brands["value_for_money"] = brands["value_for_money"].map('${:,.3f}'.format)
brands

mean_price  mean_mileage value_for_money
volkswagen          3,665       126,121          $0.029
opel                2,278       126,812          $0.018
bmw                 5,503       134,651          $0.041
mercedes_benz       9,604       115,244          $0.083
renault             1,597       122,206          $0.013
audi                5,446       141,724          $0.038

Most expensive car brand: Mercedez_benz.
Car brand with most mileage: Audi.

Audi has highest mean_mileage and 3rd most expensive car.
Mercedes is the most expensive but lowest mileage. 

The most cost-effective/value for money brand i.e. lowest price per mileage is Renault.
The least value for money brand is Mercedes. 